docs/coding-guidelines/vectorization-guidelines.md

```diff
@@ -0,0 +1,1091 @@
+public static Vector128<uint> Shuffle(Vector128<uint> vector, Vector128<uint> indices)
+public static Vector128<float> Shuffle(Vector128<float> vector, Vector128<int> indices)
+public static Vector128<long> Shuffle(Vector128<long> vector, Vector128<long> indices)
+public static Vector128<ulong> Shuffle(Vector128<ulong> vector, Vector128<ulong> indices)
+public static Vector128<double> Shuffle(Vector128<double> vector, Vector128<long> indices)
+```
+
+It can be used for many things, including reversing the input:
+
+```cs
+Vector128<int> intVector = Vector128.Create(100, 200, 300, 400);
+Console.WriteLine(intVector);
+Console.WriteLine(Vector128.Shuffle(intVector, Vector128.Create(3, 2, 1, 0)));
+```
+
+```log
+<100, 200, 300, 400>
+<400, 300, 200, 100>
+```
+
+#### Vector256.Shuffle vs Avx2.Shuffle
+
+`Vector256.Shuffle` and `Avx2.Shuffle` are not identical.
+
+`Avx2.Shuffle` is effectively `2x128-bit ops` while `Vector256.Shuffle` treats it as a "single 256-bit vector" (rather than "2x128-bit vectors"). This was done for consistency and to better map to a cross-platform mentality where `AVX-512` and `SVE` all operate on "full width".
+
+## Summary
+
+The main goal of the new `Vector128` and `Vector256` APIs is to make writing fast, vectorized code possible without becoming familiar with hardware-specific instructions and becoming an assembly language expert. Our recommendations depend on your current expertise level, software you maintain and the one you need to create:
+
+- If you are already an expert and you have vectorized your code for both `x64/x86` and `arm64/arm` code you can use the new APIs to simplify your code, but you most likely won't observe any performance gains. [#64451](https://github.com/dotnet/runtime/issues/64451) lists the places where it was/can be done in dotnet/runtime. You can use links to the merged PRs to see real-life examples.
+- If you have already vectorized your code, but only for `x64/x86` or `arm64/arm`, you can use the  new APIs to have a single, cross-platform implementation.
+- If you have already vectorized your code with `Vector<T>` you can use the new APIs to check if they can produce better codegen.
+- If you are not familiar with hardware specific instructions or you are about to vectorize a scalar algorithm, you should start with the new `Vector128` and `Vector256` APIs. Get a solid and working implementation and eventually consider using hardware-specific methods for performance critical code paths.
+
+### Best practices
+
+1. Implement tests that cover all code paths, including Access Violations.
+2. Run tests for all hardware acceleration scenarios, use the existing environment variables to do that.
+3. Implement benchmarks that mimic real life scenarios, do not increase the complexity of your code when it's not beneficial for your end users.
+4. Prefer managed references over unsafe pointers to avoid pinning and safety issues.
```

Comment:
Is this our guidance?  There are safety issues with refs as well, and there's a class of bugs that pinning helps avoid.

Reply:
From my perspective managed references are lesser evil when compared to pointers. 

@tannergooding what  is your opinion on this?

Reply:
Both are dangerous to use incorrectly and each comes with their own tradeoffs.

`pointers` are typically the safest and least problematic to get wrong because you can write natural C# syntax to do what you want. The biggest concern is that pinning for very long periods of time can negatively impact the GC, because it can't compact or otherwise move the pinned data.

`references` are greatly improved in .NET 7+ where you can use `LoadUnsafe(ref address, index)` and `StoreUnsafe(Ref address, index)` as it can help avoid the problematic classes of bugs. They allow the GC to work at its best, but that can cause additional run to run variance and can change the alignment of your data partway through.